### DISCUSSION TITLE
Input to the DAC

### DISCUSSION THREAD

**Discussion by macit1020** [pschatzmann/ESP32-A2DP#418] [Q&A] [answered]
Hi

I am quite new to programming in ES32 and I got my iphone to connect to it and have music playing to my earphone via an external 32 bit dac. For a school project I want to filter this audio. My question is how can I access the data going to the DAC. My teacher told me the data going to the dac is pcm and should be able to be filtered and need to be for both channels.

/ In the setup function:
a2dp_sink.set_stream_reader(read_data_stream);

// Then somewhere in your sketch:
void read_data_stream(const uint8_t *data, uint32_t length)

  int16_t *samples = (int16_t*) data;
  uint32_t sample_count = length/2;
  // Do something with the data packet

I saw this function in your library, but I am not sure if this function is the right. 

My final question, if I can access the data to the DAC, where can I find information to start to have the filtered data go to the dac and to my earphones

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you have a look at my [AudioTools](https://github.com/pschatzmann/arduino-audio-tools) project. The Readme and Wiki should have all the information to get you started!

It has also plenty of examples that might help you to get started. This [example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino) shows how to use the callback and just write them to the DAC. You can just replace this with a FilteredStream. There is a chapter in the Wiki about Filters!

**Reply by macit1020** [reply]
thank you for the very quick reply I will take a look through it. Should the example show something on the Serial monitor, because i saw nothing. I also saw this example: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-a2dp-serial/streams-a2dp-serial.ino.    Is the CSV data the raw data that is decrypted from my phone?
